### PR TITLE
Build against Brooklyn 0.9.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by johnmccabe**
Requires brooklyncentral/advanced-networking/pull/85.

Tested docker cloud with calico in Softlayer, deployed Node catalog app, Tomcat and redis/dnmonster/identidock examples from getting started section on http://clocker.io.

*Note* the cast on line 617 in DockerContainerImpl needs to be properly fixed.

### CODE REVIEW

**Comment by johnmccabe** [comment]
Hopefully resolves #191

**Comment by grkvlt** [comment]
Thanks @johnmccabe we'll merge this once a 1.1.0 release has been made

**Comment by grkvlt** [comment]
@johnmccabe suggest you cherry-pick commit 5eaf90ff267a702c9d9da6e322d393fd953e9c1e from #178 as well

File: docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java [hunk 1]
```diff
@@ -614,13 +614,15 @@ public DockerContainerLocation createLocation(Map flags) {
             }
 
             // Create our wrapper location around the container
-            LocationSpec<DockerContainerLocation> spec = LocationSpec.create(DockerContainerLocation.class)
+            LocationSpec<DockerContainerLocation> spec = (LocationSpec<DockerContainerLocation>) LocationSpec.create(DockerContainerLocation.class)
```
Comment (Graeme-Miller) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java]: why is this required?

Comment (Graeme-Miller) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java]: This compiles, might be neater than casting:
```java
LocationSpec<DockerContainerLocation> spec = LocationSpec.create(DockerContainerLocation.class);
            spec.parent(host)
            .configure(flags)
...
```

but still not sure what has changed to make this required

File: docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java [hunk 2]
```diff
@@ -86,12 +86,15 @@
...
-    ConfigKey<EntitySpec> SDN_PROVIDER_SPEC = ConfigKeys.newConfigKey(EntitySpec.class, "sdn.provider.spec", "SDN provider entity specification");
+    ConfigKey<EntitySpec<?>> SDN_PROVIDER_SPEC = ConfigKeys.newConfigKey(
+            new TypeToken<EntitySpec<?>>() {}, "sdn.provider.spec", "SDN provider entity specification");
 
+    @SuppressWarnings("serial")
     @SetFromFlag("hostSpec")
-    AttributeSensorAndConfigKey<EntitySpec, EntitySpec> DOCKER_HOST_SPEC = ConfigKeys.newSensorAndConfigKey(
-            EntitySpec.class, "docker.host.spec", "Specification to use when creating child Docker Hosts",
+    AttributeSensorAndConfigKey<EntitySpec<?>, EntitySpec<?>> DOCKER_HOST_SPEC = ConfigKeys.newSensorAndConfigKey(
+            new TypeToken<EntitySpec<?>>() {}, "docker.host.spec", "Specification to use when creating child Docker Hosts",
```
Comment (Graeme-Miller) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java]: as above, why is this change required?

File: docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java (hunk 1 above)
Comment (johnmccabe) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java]: Changing as suggested, didn't come up with a better way of dealing with the type erasure.

File: docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java (hunk 2 above)
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java]: Yes, I'd prefer to leave this (and similar instances) _without_ the generic types rather than explicitly saying `Blah<?>` everywhere, and having to use `TypeToken<Blah<?>>` instead of just `Blah.class` since it adds no extra information.

Comment (johnmccabe) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java]: @grkvlt I've stripped out the generics that I'd introduced from the interfaces, but have not figured out how to avoid the cast to `EntitySpec<?>` in [DockerInfrastructureImpl.java#L185](https://github.com/johnmccabe/clocker/blob/0.9.0-update/docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java#L185]) and [MesosClusterImpl.java#L149](https://github.com/johnmccabe/clocker/blob/0.9.0-update/mesos/src/main/java/brooklyn/entity/mesos/MesosClusterImpl.java#L149).

File: docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java [hunk 3]
```diff
@@ -182,7 +182,7 @@ public String apply(@Nullable Entity input) {
                 .displayName("Docker Applications"));
 
         if (config().get(SDN_ENABLE) && config().get(SDN_PROVIDER_SPEC) != null) {
-            Entity sdn = addChild(EntitySpec.create(config().get(SDN_PROVIDER_SPEC))
+            Entity sdn = addChild(EntitySpec.create((EntitySpec<?>)config().get(SDN_PROVIDER_SPEC))
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java]: We don't use the results of this cast for anything, and in fact the return type of `addChild(EntitySpec)` is just `Entity` so it should be removed

File: docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java [hunk 4]
```diff
@@ -19,51 +19,37 @@
...
-import org.apache.brooklyn.camp.brooklyn.spi.creation.BrooklynComponentTemplateResolver;
-import org.apache.brooklyn.camp.brooklyn.spi.creation.service.BrooklynServiceTypeResolver;
-import org.apache.brooklyn.camp.brooklyn.spi.creation.service.ServiceTypeResolver;
-import org.apache.brooklyn.camp.spi.PlatformComponentTemplate;
-import org.apache.brooklyn.util.text.Strings;
-
 import brooklyn.entity.container.DockerAttributes;
 
 /**
  * This converts {@link PlatformComponentTemplate} instances whose type is prefixed {@code docker:}
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java]: The Javadoc comments should be revised in line with the code changes

File: docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java [hunk 5]
```diff
@@ -71,16 +57,12 @@ public String getBrooklynType(String serviceType) {
...
         spec.configure(DockerAttributes.DOCKER_IMAGE_NAME, imageName);
         if (parts.size() == 2) {
             spec.configure(DockerAttributes.DOCKER_IMAGE_TAG, imageTag);
         }
-        if (resolver.getAttrs().containsKey("id")) {
-            String containerName = (String) resolver.getAttrs().getStringKey("id");
-            spec.configure(DockerAttributes.DOCKER_CONTAINER_NAME, containerName);
-        }
-
-        super.decorateSpec(resolver, spec);
```
Comment (grkvlt) [inline docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java]: No call to `super` needed?

File: mesos/src/main/java/brooklyn/entity/mesos/MesosClusterImpl.java [hunk 6]
```diff
@@ -146,7 +146,7 @@ public String apply(@Nullable Entity input) {
         }
 
         if (config().get(SDN_ENABLE) && config().get(SDN_PROVIDER_SPEC) != null) {
-            Entity sdn = addChild(EntitySpec.create(config().get(SDN_PROVIDER_SPEC))
+            Entity sdn = addChild(EntitySpec.create((EntitySpec<?>)config().get(SDN_PROVIDER_SPEC))
```
Comment (grkvlt) [inline mesos/src/main/java/brooklyn/entity/mesos/MesosClusterImpl.java]: See previous comment; cast not required here

**Comment by grkvlt** [comment]
https://github.com/johnmccabe/clocker/blob/0.9.0-update/docker/src/main/resources/META-INF/services/org.apache.brooklyn.core.resolve.ServiceSpecResolver

This file needs changed to point at the new `DockerEntitySpecResolver` class

File: docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java (hunk 3 above)
Comment (johnmccabe) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java]: Problem here is that without the cast what gets passed to the child is `org.apache.brooklyn.api.internal.AbstractBrooklynObjectSpec`. It's the `configure` method that's causing the problem but I don't currently see how/where to resolve it without the cast, previously `configure` was implemented in the `EntitySpec` and we didn't encounter this problem, but it now resides in `AbstractBrooklynObjectSpec` and we're losing the type along the way. I'm going to try to replicate what I'm seeing to try to understand it.

File: docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java (hunk 5 above)
Comment (johnmccabe) [inline docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java]: This is part of the cherry picked commit https://github.com/brooklyncentral/clocker/commit/5eaf90ff267a702c9d9da6e322d393fd953e9c1e from https://github.com/brooklyncentral/clocker/pull/178, I just merged and deployed a few blueprints to check, I didn't dig into the specifics. I'll have a closer look.

File: docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java (hunk 3 above)
Comment (Graeme-Miller) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java]: Similar trick to last time maybe?
```YAML
EntitySpec entitySpec = config().get(SDN_PROVIDER_SPEC);
entitySpec.configure(DockerAttributes.DOCKER_INFRASTRUCTURE, this);
Entity sdn = addChild(EntitySpec.create(entitySpec));
```

Comment (johnmccabe) [inline docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java]: Thanks @Graeme-Miller that works.

File: docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java (hunk 5 above)
Comment (neykov) [inline docker/src/main/java/brooklyn/entity/container/docker/application/DockerEntitySpecResolver.java]: That's an abstract method override, no super implementation.

**Comment by neykov** [comment]
Also worth looking through the `patches` project to remove already integrated changes.

**Comment by Graeme-Miller** [comment]
@neykov I have been looking into Clocker patches recently. Most of the patches are JClouds, and we will need a new release of JClouds to get them.

There is one patch that we should be able to get rid of, and that is JcloudsLocationSecurityGroupCustomizer (despite the name it is a brooklyn class).
